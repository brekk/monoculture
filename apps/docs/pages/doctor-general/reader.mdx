# reader



## Unknown

 @name iterateOverWorkspacesAndReadFiles  @future  @example  ```js test=true  import { fork, resolve as resolveF, parallel } from 'fluture'  // drgen-import-above  fork(done)(x => {    expect(x).toEqual([      "tools/digested",      "tools/doctor-general-cli",      "tools/gitparty",      "tools/spacework",      "tools/superorganism",      "tools/treacle",    ])    done()  })(    iterateOverWorkspacesAndReadFiles(      {        searchGlob: '*',        ignore: []      },      '../..',      resolveF(['tools/'])    )  )
