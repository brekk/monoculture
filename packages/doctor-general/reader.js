import { log } from './log'
import { chain, when, curry, flatten, map, pipe, propOr } from 'ramda'
import { parallel } from 'fluture'
import { signal } from 'kiddo'
import { readJSONFile, readDirWithConfig } from 'file-system'

export const iterateOverWorkspacesAndReadFiles = curry(
  function _iterateOverWorkspacesAndReadFiles(
    { search: searchGlob, ignore },
    root,
    x
  ) {
    log.reader('searchGlob?', searchGlob)
    return pipe(
      // look for specific file types
      map(workspace => workspace + searchGlob),
      log.core('ITERATING'),
      chain(
        readDirWithConfig({
          // exclude some search spaces
          ignore,
          cwd: root,
        })
      )
    )(x)
  }
)

export const readPackageJsonWorkspaces = curry(
  function _readPackageJsonWorkspaces(root, x) {
    return map(
      pipe(
        propOr([], 'workspaces'),
        log.parse('workspaces'),
        // we want directories only
        map(z => `${z}/`),
        map(readDirWithConfig({ cwd: root }))
      )
    )(x)
  }
)

export const monorepoRunner = curry(
  function _monorepoRunnerF(cancel, config, root, pkgJsonPath) {
    const { showMatchesOnly } = config
    log.core('CONFIG', config)
    return pipe(
      log.core('reading root package.json'),
      readJSONFile,
      readPackageJsonWorkspaces(root),
      chain(parallel(10)),
      signal(cancel, {
        text: 'Reading all workspaces...',
        successText: 'Read all workspaces!',
        failText: 'Unable to read all workspaces.',
      }),
      map(
        pipe(
          flatten,
          log.core('workflows, flat'),
          iterateOverWorkspacesAndReadFiles(config, root)
        )
      ),
      chain(parallel(10)),
      when(
        () => !showMatchesOnly,
        pipe(
          map(flatten),
          log.verbose('all files read'),
          signal(cancel, {
            text: 'Reading all files...',
            successText: 'Read all files!',
            failText: 'Unable to read all files.',
          })
        )
      )
    )(pkgJsonPath)
  }
)
