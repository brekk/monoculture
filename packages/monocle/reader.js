import path from 'node:path'

import {
  identity as I,
  curry,
  pipe,
  map,
  trim,
  split,
  addIndex,
  chain,
} from 'ramda'
import { parallel } from 'fluture'

import { readDirWithConfig, readFile } from 'file-system'
import { futureFileProcessor } from 'monorail'

import { hash } from './hash'
import { log } from './log'

export const readMonoFile = curry(
  function _readMonoFile(basePath, trimContent, file) {
    return pipe(
      log.file('reading'),
      readFile,
      map(buf =>
        pipe(
          split('\n'),
          // ostensibly files have content that starts at a 1 index
          addIndex(map)((y, i) => [i + 1, trimContent ? trim(y) : y]),
          body => ({
            name: path.relative(basePath, file),
            hash: hash(buf),
            body,
          })
        )(buf)
      )
    )(file)
  }
)

export const readAll = curry(function _readAll(config, dirglob) {
  return pipe(
    log.file('reading glob'),
    readDirWithConfig({ ...config, nodir: true }),
    config.showMatchesOnly
      ? I
      : chain(files =>
          pipe(map(readMonoFile(config.cwd, config.trim)), parallel(10))(files)
        )
  )(dirglob)
})

export const monoprocessor = curry(
  function _monoprocessor(config, pluginsF, dirGlob) {
    return pipe(
      readAll(config),
      config.showMatchesOnly ? I : futureFileProcessor(config, pluginsF)
    )(dirGlob)
  }
)
