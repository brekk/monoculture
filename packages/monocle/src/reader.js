import path from 'node:path'

import {
  identity as I,
  curry,
  pipe,
  map,
  trim,
  split,
  addIndex,
  chain,
} from 'ramda'
import { parallel } from 'fluture'

import { readDirWithConfig, readFile } from 'file-system'
import { futureFileProcessor } from 'monorail'

import { hash } from './hash'
import { log } from './trace'

export const readMonoFile = curry((basePath, trimContent, file) =>
  pipe(
    log.file('reading'),
    readFile,
    map(buf =>
      pipe(
        split('\n'),
        // ostensibly files have content that starts at a 1 index
        addIndex(map)((y, i) => [i + 1, trimContent ? trim(y) : y]),
        body => ({
          name: path.relative(basePath, file),
          hash: hash(buf),
          body,
        })
      )(buf)
    )
  )(file)
)

export const readAll = curry((config, dirglob) => {
  return pipe(
    log.file('reading glob'),
    readDirWithConfig({ ...config, nodir: true }),
    config.showMatchesOnly
      ? I
      : chain(files =>
          pipe(map(readMonoFile(config.cwd, config.trim)), parallel(10))(files)
        )
  )(dirglob)
})

export const monoprocessor = curry((config, pluginsF, dirGlob) =>
  pipe(
    readAll(config),
    config.showMatchesOnly ? I : futureFileProcessor(config, pluginsF)
  )(dirGlob)
)
