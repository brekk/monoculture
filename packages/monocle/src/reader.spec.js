import { readMonoFile, readAll, monoprocessor } from './reader'
import { resolve, fork } from 'fluture'

test('readMonoFile', done => {
  expect(readMonoFile).toBeTruthy()
  const readF = readMonoFile('zip/zap/zop', true, './src/cli.js')
  fork(done)(x => {
    expect(x.file).toEqual('../../../src/cli.js')
    expect(x.hash).toMatchSnapshot()
    done()
  })(readF)
})
const DIRFILES = [
  './src/cli.js',
  './src/config.js',
  './src/config.spec.js',
  './src/hash.js',
  './src/hash.spec.js',
  './src/index.js',
  './src/index.spec.js',
  './src/reader.js',
  './src/reader.spec.js',
  './src/trace.js',
]
test('readAll', done => {
  expect(readAll).toBeTruthy()
  const readDF = readAll({ showMatchesOnly: true }, './src/*')
  fork(done)(x => {
    expect(x).toEqual(DIRFILES)
    done()
  })(readDF)
})
test('monoprocessor', done => {
  expect(monoprocessor).toBeTruthy()
  const monoF = monoprocessor(
    { showMatchesOnly: true, trim: false, basePath: __dirname },
    resolve([]),
    './src/*'
  )
  fork(done)(z => {
    expect(z).toEqual(DIRFILES)
    done()
  })(monoF)
})
