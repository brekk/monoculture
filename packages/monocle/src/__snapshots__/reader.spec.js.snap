// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readMonoFile 1`] = `"b06982b10f9b42aac49f5422f41e390843cc359e823363b45d8c7a72a3005d5a"`;
